File: content-solutions/implementation/sim-date-add.py
def solution(date, days):
    def is_leap(y):
        return (y % 4 == 0 and y % 100 != 0) or y % 400 == 0

    def month_days(y, m):
        if m == 2:
            return 29 if is_leap(y) else 28
        return 30 if m in (4, 6, 9, 11) else 31

    y, m, d = map(int, date.split("-"))
    day = d - 1 + sum(month_days(y, k) for k in range(1, m)) + days
    while day >= (366 if is_leap(y) else 365):
        day -= 366 if is_leap(y) else 365
        y += 1
    m = 1
    while day >= month_days(y, m):
        day -= month_days(y, m)
        m += 1
    return f"{y:04d}-{m:02d}-{day + 1:02d}"

File: content-solutions/implementation/sim-date-add.java
import java.util.*;

class Solution {
    boolean isLeap(int y) {
        return (y % 4 == 0 && y % 100 != 0) || y % 400 == 0;
    }

    int monthDays(int y, int m) {
        if (m == 2) return isLeap(y) ? 29 : 28;
        return (m == 4 || m == 6 || m == 9 || m == 11) ? 30 : 31;
    }

    public String solution(String date, int days) {
        String[] p = date.split("-");
        int y = Integer.parseInt(p[0]), m = Integer.parseInt(p[1]), d = Integer.parseInt(p[2]);
        long day = d - 1 + (long) days;
        for (int k = 1; k < m; k++) day += monthDays(y, k);
        while (day >= (isLeap(y) ? 366 : 365)) {
            day -= isLeap(y) ? 366 : 365;
            y++;
        }
        m = 1;
        while (day >= monthDays(y, m)) {
            day -= monthDays(y, m);
            m++;
        }
        return String.format("%04d-%02d-%02d", y, m, day + 1);
    }
}
